draft-xie-bidirectional-messaging-latest.md

```diff
@@ -30,63 +30,74 @@ author:
-are framed more than once at different protocol layers, which offsets the wire
-efficiency from HTTP/2 binary frame. Take WebSocket over HTTP/2 as an example,
-user data is framed at application protocol, WebSocket, and HTTP/2 layers. This
-not only introduces the byte overhead on the wire, but also complicates the data
-processing. These techniques also pose new operational challenges to
-intermediaries. Due to the traffic from the whole user session is encapsulated
-into one HTTP/2 stream, this stream can last very long time. Intermediaries may
-take long time to drain these streams. As a remind, GOAWAY can only signals the
-remote endpoint to stop using the connection for new streams and does not help
-the open streams. Moreover, intermediaries have no visibility to user
-interactions, and lose the capability to collect telemetry metrics (e.g., time
-to the first/last byte of request and response) for services.
+to clients, but clients cannot respond to them and either accept or discard
+them silently. Additionally, intermediaries along the path may have different
+server push policies and may not forward push promised streams to the downstream
+client. This best effort mechanism is not sufficient to reliably deliver content
+from servers to clients, limiting additional use-cases, such as sending messages
+and notifications from servers to clients immediately when they become
+available.
+
+To work around this limitation, several techniques are used, like long polling
+{{!RFC6202}}, WebSocket {{!RFC8441}}, and tunneling. These techniques generally
+layer application protocols on top of HTTP/2 and use HTTP/2 streams as transport
+connections. These solutions often defeat optimizations provided by HTTP/2.
+Multiplexing multiple parallel interactions onto one HTTP/2 stream reintroduces
+head of line blocking within that stream, application metadata is often
+encapsulated into DATA frames rather than HEADERS frames, so header compression
+is no longer effective, and adding an additional layer of framing to the user
+data reduces the wire efficiency advantages of HTTP/2 binary framing.
+
+For example, WebSocket over HTTP/2 frames user data both at the WebSocket layer
+and the HTTP/2 layer. This not only introduces additional overhead on the wire,
+but also complicates the data processing. These techniques also pose new
+operational challenges to intermediaries. If traffic from an entire user session
+is encapsulated into one HTTP/2 stream, this stream can last a very long time.
+Intermediaries may take long time to drain these streams. Unfortunately, GOAWAY
+only signals the remote endpoint to stop using the connection for new streams
+and often does not help drain the remaining open streams. Moreover,
+intermediaries have no visibility to user interactions, and lose the capability
+to collect telemetry metrics (e.g., time to the first/last byte of request and
+response) for services.
```

Comment:
It seemed to me like there was additional value here for intermediaries to be able to properly route everything, perhaps we should replace this sentence with something that focuses on them being able to help out better rather than taking telemetry?

Reply:
Moving these to issues since we can tackle them later.